File: Java/No3LongestSubstringWithoutRepeatingCharacters.java
import java.util.Map;

class No3LongestSubstringWithoutRepeatingCharacters {
  public int lengthOfLongestSubstring(String s) {
    Map<Character, Integer> lastOccur = new HashMap();
    int maxLen = 0;
    int left = -1;
    for (int i = 0; i < s.length(); i++) {
      char c = s.charAt(i);
      if (lastOccur.containsKey(c) && lastOccur.get(c) >= left) {
        left = lastOccur.get(c);
      }
      lastOccur.put(c, i);
      maxLen = Math.max(maxLen, i - left);
    }          
    return maxLen;
  }
}

File: Kotlin/No3LongestSubstringWithoutRepeatingCharacters.kt
fun lengthOfLongestSubstring(s: String): Int {
    val lastOccur = mapOf<Char, Int>()
    var maxLength = 0
    var left = -1
    s.toCharArray().forEachIndexed{ i,c -> 
      lastOccur[c]?.let {
        left = if (it >= left) it else left
      }
      lastOccur[c] = i
      maxLength = if (i-left > maxLength) i-left else maxLength
    }
    return maxLength
}